SignalServiceKit/src/Messages/OWSMessageReceiver.m

```diff
@@ -0,0 +1,338 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+#import "OWSMessageReceiver.h"
+#import "OWSSignalServiceProtos.pb.h"
+#import "TSDatabaseView.h"
+#import "TSMessagesManager.h"
+#import "TSStorageManager.h"
+#import "TSYapDatabaseObject.h"
+#import <YapDatabase/YapDatabaseConnection.h>
+#import <YapDatabase/YapDatabaseTransaction.h>
+#import <YapDatabase/YapDatabaseView.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+#pragma mark - Persisted data model
+@class OWSSignalServiceProtosEnvelope;
+
+@interface OWSMessageProcessingJob : TSYapDatabaseObject
+
+@property (nonatomic, readonly) OWSSignalServiceProtosEnvelope *envelopeProto;
+@property (nonatomic, readonly) NSDate *createdAt;
+
+- (instancetype)initWithEnvelope:(OWSSignalServiceProtosEnvelope *)envelope NS_DESIGNATED_INITIALIZER;
+- (instancetype)initWithUniqueId:(NSString *)uniqueId NS_UNAVAILABLE;
+
+@end
+
+@interface OWSMessageProcessingJob ()
+
+@property (nonatomic, readonly) NSData *envelopeData;
+
+@end
+
+// TODO rename?
+@implementation OWSMessageProcessingJob
+
+- (instancetype)initWithEnvelope:(OWSSignalServiceProtosEnvelope *)envelope
+{
+    self = [super initWithUniqueId:[NSUUID new].UUIDString];
+    if (!self) {
+        return self;
+    }
+
+    _envelopeData = envelope.data;
+    _createdAt = [NSDate new];
+
+    return self;
+}
+
+- (OWSSignalServiceProtosEnvelope *)envelopeProto
+{
+    return [OWSSignalServiceProtosEnvelope parseFromData:self.envelopeData];
```

Comment:
If we received an invalid envelope, would this return null?  Should this have a nullable return value?